src/components/FallenHeroes.tsx

```diff
@@ -11,32 +11,46 @@ function FallenHeroes() {
   // matching hero from the set of totalHeroes,
   // and is included in rest of heroes alive, update the state accordingly
   const { campaign, setCampaign } = React.useContext(LotrContext);
+
+  function filterFallen() {
+    return campaign.allHeroes.filter((hero) => hero.alive === false);
+  }
+
+  const fallen = filterFallen();
   const [fallenHero, setFallenHero] = useState("");
 
+  const names = campaign.allHeroes.map((hero) => {
+    return hero.name;
+  });
+
+  const isAlive =
+    campaign.allHeroes.find((hero) => hero.name === fallenHero)?.alive === true;
+
   function handleChange(e: React.ChangeEvent<HTMLInputElement>) {
     setFallenHero(e.target.value);
   }
 
   function addFallenHero(e: React.FormEvent<HTMLFormElement>) {
     e.preventDefault();
     //add hero to the list if it isnt there yet
-    if (!campaign.restOfAliveHeroes.includes(fallenHero)) {
+    if (!names.includes(fallenHero)) {
+      throw Error("Sorry, your hero is an unknown!");
+    }
+
+    if (names.includes(fallenHero) && !isAlive) {
       throw Error("Sorry, your hero is not among the living!");
     }
-    if (campaign.restOfAliveHeroes.includes(fallenHero)) {
-      const updatedFallen = [...campaign.fallenHeroes, fallenHero];
-      const updatedRestOfAlive = campaign.restOfAliveHeroes.filter(
-        (hero) => hero !== fallenHero
-      );
-      setCampaign({
-        ...campaign,
-        fallenHeroes: updatedFallen,
-        restOfAliveHeroes: updatedRestOfAlive,
+
+    if (names.includes(fallenHero) && isAlive) {
+      campaign.allHeroes.find((hero) => {
+        if (hero.name === fallenHero) {
+          hero.alive = false;
+          setCampaign({ ...campaign });
```

Comment:
how come this does not just mutate it as well?

Reply:
Because this makes a copy of the original object, it does not point to the original object anymore. Here is a quick explanation: https://www.youtube.com/watch?v=gOl6rU1pkL0

Reply:
I see